### ISSUE TITLE
memory leak in the pure python Unpacker (when using .feed())

### ISSUE THREAD

**Issue by ThomasWaldmann** [msgpack/msgpack-python#283]
see there: https://github.com/borgbackup/borg/issues/3573

i had a quick look at the code and somehow i only found places where it grows self._buffer (e.g. in Unpacker.feed), but no place where it shrinks it. did i overlook something?

**Comment by jfolz** [comment]
The `Unpacker` will attempt to clear finished data from its buffer before trying to read: [fallback.py#L328](https://github.com/msgpack/msgpack-python/blob/4b72b6177321ef24b6c7af2354fb980db69e2aec/msgpack/fallback.py#L328). Otherwise the 
The `Packer` creates a new buffer after every call to `pack` if `autoreset=True` (the default): [fallback.py#L857](https://github.com/msgpack/msgpack-python/blob/4b72b6177321ef24b6c7af2354fb980db69e2aec/msgpack/fallback.py#L857)

**Comment by jfolz** [comment]
The strategy used to grow the buffer might differ though, and clearing the buffer might happen at different times.

**Comment by ThomasWaldmann** [comment]
Hmm, I've just closed the borgbackup issue as the user who had the problem (with pure python msgpack) said he does not have the problem with C msgpack.

**Comment by jfolz** [comment]
On second thought `Unpacker` only clears its buffer when it needs to read from stream. This never happens in feeding mode, in fact it results in an exception being thrown. We should probably look into clearing the buffer in `feed()` if necessary. This is what is done in Cython code.
For the time being I suggest using `unpackb` as a workaround if possible. It's actually much faster in Cython than reusing the same `Unpacker` instance, since it treats the memory you throw at it as a fixed buffer.

**Comment by jfolz** [comment]
Small test that this is what's happening:
```
from __future__ import print_function, division, unicode_literals
import io
import msgpack
from msgpack.fallback import *


print(msgpack.version)
p = Packer()
m = p.pack([u'hello']*10000)


bio = io.BytesIO(m)
u = Unpacker(bio)
for i in range(1000):
    bio.seek(0)
    u.unpack()
    print('stream', i, len(u._buffer) / 1e6)


u = Unpacker()
for i in range(1000):
    u.feed(m)
    u.unpack()
    print('feed', i, len(u._buffer) / 1e6)
```

Buffer length stays constant when streaming and increases with every call to feed when feeding.
Observing memory usage of the interpreter also confirms that cython code does not suffer from buffer growth.

**Comment by ThomasWaldmann** [comment]
Thanks for reproducing / analysing the issue.

**Comment by ThomasWaldmann** [comment]
Looks like the leak was fixed with the 0.5.5 release, but which release introduced the memory leak?

**Comment by methane** [comment]
I don't know.  Maybe, when fallback.py added.
